### PR TITLE
bypass permissionless concentrated liquidity checks

### PR DESCRIPTION
**Pull request by sunnya97**
Closes: #6345

Creates a governance set whitelist of addresses that can bypass the normal pool creation restrictions on concentrated liquidity pools

## What is the purpose of the change

- Allow a governance subdao automatically approved stAsset/Asset pools for non-approved quote assets.
- Allow a governance subdao to create pools when permissionless pool creation is off

## Testing and Verifying

  - Added unit tests

## Documentation and Release Note

  - [ ] Does this pull request introduce a new feature or user-facing behavior changes?
  - [ ] Changelog entry added to `Unreleased` section of `CHANGELOG.md`?

Where is the change documented? 
  - [ ] Specification (`x/{module}/README.md`)
  - [ ] Osmosis documentation site
  - [ ] Code comments?
  - [ ] N/A

### CODE REVIEW

File: x/concentrated-liquidity/pool_test.go [hunk 1]
```diff
@@ -91,6 +111,16 @@ func (s *KeeperTestSuite) TestInitializePool() {
...
 			}
 
+			if len(test.unrestrictedPoolCreatorWhitelist) > 0 {
+				params := s.App.ConcentratedLiquidityKeeper.GetParams(s.Ctx)
+				params.UnrestrictedPoolCreatorWhitelist = test.unrestrictedPoolCreatorWhitelist
+				fmt.Println("setting")
+				fmt.Println(params.UnrestrictedPoolCreatorWhitelist)
+				s.App.ConcentratedLiquidityKeeper.SetParams(s.Ctx, params)
+				fmt.Println("getting")
+				fmt.Println(s.App.ConcentratedLiquidityKeeper.GetParams(s.Ctx).UnrestrictedPoolCreatorWhitelist)
```
Comment (p0mvn) [inline x/concentrated-liquidity/pool_test.go:121]: nit: prints but I saw them being removed in a follow-up PR so good with keeping them until then

**Review by p0mvn — APPROVED** [review]
LGTM once CI is happy!

Need to add a changelog entry for everything to pass

**Review by mattverse — APPROVED** [review]
Implementation straight forward and lgtm for me as well, I'm wondering why e2e is failing, Im wondering if it could be the new param that we are adding, might need to handle it in upgrades.go?

File: x/concentrated-liquidity/pool_test.go [hunk 2]
```diff
@@ -91,6 +111,16 @@ func (s *KeeperTestSuite) TestInitializePool() {
 				s.App.PoolManagerKeeper.SetParams(s.Ctx, params)
 			}
 
+			if len(test.unrestrictedPoolCreatorWhitelist) > 0 {
+				params := s.App.ConcentratedLiquidityKeeper.GetParams(s.Ctx)
+				params.UnrestrictedPoolCreatorWhitelist = test.unrestrictedPoolCreatorWhitelist
+				fmt.Println("setting")
+				fmt.Println(params.UnrestrictedPoolCreatorWhitelist)
```
Comment (sunnya97) [inline x/concentrated-liquidity/pool_test.go:118]:
```suggestion
```

File: x/concentrated-liquidity/pool_test.go (hunk 1 above)
Comment (sunnya97) [inline x/concentrated-liquidity/pool_test.go:121]:
```suggestion
```

File: CHANGELOG.md [hunk 3]
```diff
@@ -1018,7 +1019,7 @@ Upgrade instructions for node operators can be found [here](https://github.com/o
 * Make queries to bank total chain balance account for developer vesting correctly.
 * Add ability for nodes to query the total amount locked for each denomination.
 * Embedded seeds in init.go
-* Added changelog and info about changelog format.
+* Added changelog and info about  format.
```
Comment (sunnya97) [inline CHANGELOG.md:1022]:
```suggestion
* Added changelog and info about changelog format.
```

**Comment by sunnya97** [comment]
I'm not sure why my e2e is failing ☹️

**Comment by p0mvn** [comment]
> I'm not sure why my e2e is failing ☹️

There was an issue on `main`. Got fixed yesterday by: https://github.com/osmosis-labs/osmosis/pull/6426

Merging `main` here might help

**Comment by czarcas7ic** [comment]
Its likely not an E2E issue, it looks like its failing for a legitimate issue. Will need to add it to my list and will get to this if not figured out

File: proto/osmosis/concentrated-liquidity/params.proto [hunk 4]
```diff
@@ -56,4 +56,11 @@ message Params {
   bool is_permissionless_pool_creation_enabled = 6
       [ (gogoproto.moretags) =
             "yaml:\"is_permissionless_pool_creation_enabled\"" ];
+
+  // unrestricted_pool_creator_whitelist is a list of addresses that are
+  // allowed to bypass restrictions on permissionless supercharged pool
+  // creation, like pool_creation_enabled, restricted quote assets, no
+  // double creation of pools, etc.
+  repeated string unrestricted_pool_creator_whitelist = 7
+      [ (gogoproto.moretags) = "yaml:\"unrestricted_pool_creator_whitelist\"" ];
```
Comment (p0mvn) [inline proto/osmosis/concentrated-liquidity/params.proto:65]: I suspect that we might need to initialize this to something in the upgrade handler (can be an empty slice)

Comment (p0mvn) [reply proto/osmosis/concentrated-liquidity/params.proto:65]: Will see if this passes: https://github.com/osmosis-labs/osmosis/pull/6451

Comment (p0mvn) [reply proto/osmosis/concentrated-liquidity/params.proto:65]: Nope that did not work

**Comment by czarcas7ic** [comment]
I see the problem, we are failing in both tests that call MsgSetMaxPoolPointsPerTx with the following panic:

```        
raw_log: "recovered: UnmarshalJSON cannot decode empty bytes\nstack:\ngoroutine 178
          [running]:\nruntime/debug.Stack()\n\truntime/debug/stack.go:24 +0x64\ngithub.com/cosmos/cosmos-sdk/baseapp.newDefaultRecoveryMiddleware.func1({0x231a320,
          0x400649f960})\n\tgithub.com/cosmos/cosmos-sdk@v0.47.5/baseapp/recovery.go:71 +0x24\ngithub.com/cosmos/cosmos-sdk/baseapp.newRecoveryMiddleware.func1({0x231a320?,
          0x400649f960?})\n\tgithub.com/cosmos/cosmos-sdk@v0.47.5/baseapp/recovery.go:39 +0x34\ngithub.com/cosmos/cosmos-sdk/baseapp.processRecovery({0x231a320,
          0x400649f960}, 0x400000e270?)\n\tgithub.com/cosmos/cosmos-sdk@v0.47.5/baseapp/recovery.go:28
          +0x38\ngithub.com/cosmos/cosmos-sdk/baseapp.processRecovery({0x231a320, 0x400649f960},
          0x2df5560?)\n\tgithub.com/cosmos/cosmos-sdk@v0.47.5/baseapp/recovery.go:33 +0x60\ngithub.com/cosmos/cosmos-sdk/baseapp.(*BaseApp).runTx.func1()\n\tgithub.com/cosmos/cosmos-sdk@v0.47.5/baseapp/baseapp.go:620
          +0xa8\npanic({0x231a320, 0x400649f960})\n\truntime/panic.go:884 +0x1f4\ngithub.com/cosmos/cosmos-sdk/x/params/types.Subspace.Get({{_,
          _}, _, {_, _}, {_, _}, {_, _, _}, ...}, ...)\n\tgithub.com/cosmos/cosmos-sdk@v0.47.5/x/params/types/subspace.go:109
          +0x20c\ngithub.com/cosmos/cosmos-sdk/x/params/types.Subspace.GetParamSet({{_, _},
          _, {_, _}, {_, _}, {_, _, _}, ...}, ...)\n\tgithub.com/cosmos/cosmos-sdk@v0.47.5/x/params/types/subspace.go:222
          +0x100\ngithub.com/osmosis-labs/osmosis/v19/x/concentrated-liquidity.Keeper.GetParams(...)\n\tgithub.com/osmosis-labs/osmosis/v19/x/concentrated-liquidity/keeper.go:53\ngithub.com/osmosis-labs/osmosis/v19/x/concentrated-liquidity.Keeper.ValidatePermissionlessPoolCreationEnabled(...)\n\tgithub.com/osmosis-labs/osmosis/v19/x/concentrated-liquidity/keeper.go:124\ngithub.com/osmosis-labs/osmosis/v19/x/poolmanager.Keeper.CreatePool({{0x2dc0848,
          0x4000f6ec10}, {0x2df4688, 0x4000142600}, {0x2df4598, 0x4000015320}, {0x2df4610,
          0x400070dc70}, {0x2db88a0, 0x40005ee0e0}, ...}, ...)\n\tgithub.com/osmosis-labs/osmosis/v19/x/poolmanager/create_pool.go:41
          +0x7c\ngithub.com/osmosis-labs/osmosis/v19/x/concentrated-liquidity.msgServer.CreateConcentratedPool({0x238f940?},
          {0x2ddff18?, 0x40046c0180?}, 0xdfa4d0?)\n\tgithub.com/osmosis-labs/osmosis/v19/x/concentrated-liquidity/msg_server.go:39
          +0xcc\ngithub.com/osmosis-labs/osmosis/v19/x/concentrated-liquidity/model._Msg_CreateConcentratedPool_Handler.func1({0x2ddff18,
          0x40046c0180}, {0x2611620?, 0x4004367300})\n\tgithub.com/osmosis-labs/osmosis/v19/x/concentrated-liquidity/model/tx.pb.go:252
          +0x74\ngithub.com/cosmos/cosmos-sdk/baseapp.(*MsgServiceRouter).RegisterService.func2.1({0x2ddff18,
          0x40046c0150}, {0x40017f43a8?, 0xdfa2c4?}, 0x264d700?, 0x4006598c78)\n\tgithub.com/cosmos/cosmos-sdk@v0.47.5/baseapp/msg_service_router.go:113
          +0x98\ngithub.com/osmosis-labs/osmosis/v19/x/concentrated-liquidity/model._Msg_CreateConcentratedPool_Handler({0x2471640?,
          0x4000126578}, {0x2ddff18, 0x40046c0150}, 0x2b23b20, 0x40046b47c0)\n\tgithub.com/osmosis-labs/osmosis/v19/x/concentrated-liquidity/model/tx.pb.go:254
          +0x12c\ngithub.com/cosmos/cosmos-sdk/baseapp.(*MsgServiceRouter).RegisterService.func2({{0x2ddff18,
          0x4004613d70}, {0x2df5560, 0x40046b0840}, {{0xb, 0x1}, {0x4006b873e0, 0xb}, 0x5d,
          {0x228e54ab, ...}, ...}, ...}, ...)\n\tgithub.com/cosmos/cosmos-sdk@v0.47.5/baseapp/msg_service_router.go:118
          +0x288\ngithub.com/cosmos/cosmos-sdk/baseapp.(*BaseApp).runMsgs(_, {{0x2ddff18,
          0x4004613d70}, {0x2df5560, 0x40046b0840}, {{0xb, 0x1}, {0x4006b873e0, 0xb}, 0x5d,
          ...}, ...}, ...)\n\tgithub.com/cosmos/cosmos-sdk@v0.47.5/baseapp/baseapp.go:763
          +0x1ec\ngithub.com/cosmos/cosmos-sdk/baseapp.(*BaseApp).runTx(0x400140e900, 0x3,
          {0x4004552180, 0x169, 0x169})\n\tgithub.com/cosmos/cosmos-sdk@v0.47.5/baseapp/baseapp.go:705
          +0x948\ngithub.com/cosmos/cosmos-sdk/baseapp.(*BaseApp).DeliverTx(0x400140e900,
          {{0x4004552180?, 0x400663e300?, 0x20?}})\n\tgithub.com/cosmos/cosmos-sdk@v0.47.5/baseapp/abci.go:308
          +0xf4\ngithub.com/tendermint/tendermint/abci/client.(*localClient).DeliverTxAsync(0x40015c6e40,
          {{0x4004552180?, 0x40015c6e40?, 0xeda5d8?}})\n\tgithub.com/tendermint/tendermint@v0.37.0-rc1/abci/client/local_client.go:93
          +0xec\ngithub.com/tendermint/tendermint/proxy.(*appConnConsensus).DeliverTxAsync(0x4004c29e20?,
          {{0x4004552180?, 0x20?, 0xb?}})\n\tgithub.com/tendermint/tendermint@v0.37.0-rc1/proxy/app_conn.go:85
          +0x2c\ngithub.com/tendermint/tendermint/state.execBlockOnProxyApp({0x2de1488?, 0x40012ded80},
          {0x2dea7d0, 0x400116b8a0}, 0x4001b625a0, {0x2df49d0, 0x40012a9560}, 0x5c?)\n\tgithub.com/tendermint/tendermint@v0.37.0-rc1/state/execution.go:320
          +0x520\ngithub.com/tendermint/tendermint/state.(*BlockExecutor).ApplyBlock(_, {{{0xb,
          0x1}, {0x40000dcc80, 0x7}}, {0x40000dcc90, 0xb}, 0x1, 0x5c, {{0x400629f6a0, ...},
          ...}, ...}, ...)\n\tgithub.com/tendermint/tendermint@v0.37.0-rc1/state/execution.go:140
          +0xf0\ngithub.com/tendermint/tendermint/consensus.(*State).finalizeCommit(0x40013ff880,
          0x5d)\n\tgithub.com/tendermint/tendermint@v0.37.0-rc1/consensus/state.go:1654 +0x958\ngithub.com/tendermint/tendermint/consensus.(*State).tryFinalizeCommit(0x40013ff880,
          0x5d)\n\tgithub.com/tendermint/tendermint@v0.37.0-rc1/consensus/state.go:1563 +0x278\ngithub.com/tendermint/tendermint/consensus.(*State).enterCommit.func1()\n\tgithub.com/tendermint/tendermint@v0.37.0-rc1/consensus/state.go:1498
          +0x8c\ngithub.com/tendermint/tendermint/consensus.(*State).enterCommit(0x40013ff880,
          0x5d, 0x0)\n\tgithub.com/tendermint/tendermint@v0.37.0-rc1/consensus/state.go:1536
          +0xb48\ngithub.com/tendermint/tendermint/consensus.(*State).addVote(0x40013ff880,
          0x4004336500, {0x40004e2c30, 0x28})\n\tgithub.com/tendermint/tendermint@v0.37.0-rc1/consensus/state.go:2150
          +0x15f4\ngithub.com/tendermint/tendermint/consensus.(*State).tryAddVote(0x40013ff880,
          0x4004336500, {0x40004e2c30?, 0xf4f054?})\n\tgithub.com/tendermint/tendermint@v0.37.0-rc1/consensus/state.go:1948
          +0x28\ngithub.com/tendermint/tendermint/consensus.(*State).handleMsg(0x40013ff880,
          {{0x2db90c0, 0x40018158d0}, {0x40004e2c30, 0x28}})\n\tgithub.com/tendermint/tendermint@v0.37.0-rc1/consensus/state.go:853
          +0x364\ngithub.com/tendermint/tendermint/consensus.(*State).receiveRoutine(0x40013ff880,
          0x0)\n\tgithub.com/tendermint/tendermint@v0.37.0-rc1/consensus/state.go:760 +0x2b0\ncreated
          by github.com/tendermint/tendermint/consensus.(*State).OnStart\n\tgithub.com/tendermint/tendermint@v0.37.0-rc1/consensus/state.go:379
          +0xf0\n: panic"
        timestamp: ""
        tx: null
        txhash: BE5B7CEF46EBDDD1E0E7CE45EAAE500D91C055882F95ACAF0C2E133913AE1D26
```

need to look into why now, just posting the log for anyone following along

**Comment by czarcas7ic** [comment]
So I fixed it for this PR, but I realized what the issue is that still needs to be addressed:

1. We add a new param to a module
2. We attempt to GetParams in the upgrade handler
3. It errors because we dont re-register the ParamKeyTable if it already exists https://github.com/osmosis-labs/osmosis/blob/f219b45d38436ec68151869968ed2e65f3a388e1/x/concentrated-liquidity/keeper.go#L32-L36

So I think the solution is if we add a new param, we need to regen this ParamKeyTable despite it already existing. This is exactly the reason we got the mainnet bug where pool creation fee got set to 1000 OSMO. I had to work around this using DefaultParams but didn't set that value back.

**Comment by czarcas7ic** [comment]
I better understand the issue now and have the full fix, I will commit to this PR since its fairly related

**Comment by czarcas7ic** [comment]
Re-requesting reviews since I added logic on top of this, but logic that is required to properly initialize new params for this PR and future PRs.

https://github.com/osmosis-labs/osmosis/pull/6420/commits/0441ca3f66df44fa619386f46a6422582b13895a

File: app/upgrades/v20/upgrades.go [hunk 5]
```diff
@@ -23,6 +24,9 @@ func CreateUpgradeHandler(
 			return nil, err
 		}
 
+		// Initialize the newly created param
+		keepers.ConcentratedLiquidityKeeper.SetParam(ctx, cltypes.KeyUnrestrictedPoolCreatorWhitelist, []string{})
```
Comment (p0mvn) [inline app/upgrades/v20/upgrades.go:28]: This feels like something that the upstream SDK should automate. WDYT about opening up an issue?

**Review by p0mvn — APPROVED** [review]
e2e fix makes sense to me! Nice work @czarcas7ic and @sunnya97

File: app/upgrades/v20/upgrades.go (hunk 5 above)
Comment (czarcas7ic) [reply app/upgrades/v20/upgrades.go:28]: I will open up an issue rn and link it here, good idea

Comment (czarcas7ic) [reply app/upgrades/v20/upgrades.go:28]: Created https://github.com/cosmos/cosmos-sdk/issues/17826#issue-1905750219

**Comment by sunnya97** [comment]
Ah yes, this makes sense

**Review by mattverse — APPROVED** [review]
Nice work on the investigation @czarcas7ic ! 🎊 

IMO we should have an API for this upstream where we can pass in the param we are adding and then rest is taken care of in sdk

**Review by czarcas7ic — APPROVED** [review]
Approved, merging, thanks all

**Pull request merged**